lib/charms/s3_integrator/v0/s3.py

```diff
@@ -0,0 +1,479 @@
+        only the leader unit can call it.
+
+        Args:
+            relation_id: the identifier for a particular relation.
+            attributes: the attributes value.
+        """
+        self.update_credential_data(relation_id, {"attributes": attributes})
+
+
+class S3Event(RelationEvent):
+    """Base class for S3 storage events."""
+
+    @property
+    def bucket(self) -> Optional[str]:
+        """Returns the bucket name."""
+        return self.relation.data[self.relation.app].get("bucket", default=None)
+
+    @property
+    def access_key(self) -> Optional[str]:
+        """Returns the access key."""
+        return self.relation.data[self.relation.app].get("access-key", default=None)
+
+    @property
+    def secret_key(self) -> Optional[str]:
+        """Returns the secret key."""
+        return self.relation.data[self.relation.app].get("secret-key", default=None)
+
+    @property
+    def path(self) -> Optional[str]:
+        """Returns the path where data can be stored."""
+        return self.relation.data[self.relation.app].get("path", default=None)
+
+    @property
+    def endpoint(self) -> Optional[str]:
+        """Returns the enpoint address."""
+        return self.relation.data[self.relation.app].get("endpoint", default=None)
+
+    @property
+    def region(self) -> Optional[str]:
+        """Returns the region."""
+        return self.relation.data[self.relation.app].get("region", default=None)
+
+    @property
+    def s3_uri_style(self) -> Optional[str]:
+        """Returns the s3 uri style."""
+        return self.relation.data[self.relation.app].get("s3-uri-style", default=None)
+
+    @property
+    def storage_class(self) -> Optional[str]:
+        """Returns the storage class name."""
+        return self.relation.data[self.relation.app].get("storage-class", default=None)
+
+    @property
+    def tls_ca_chain(self) -> Optional[str]:
+        """Returns the TLS CA chain."""
+        return self.relation.data[self.relation.app].get("tls-ca-chain", default=None)
+
+    @property
+    def s3_api_version(self) -> Optional[str]:
+        """Returns the S3 API version."""
+        return self.relation.data[self.relation.app].get("s3-api-version", default=None)
+
+    @property
+    def attributes(self) -> Optional[str]:
+        """Returns the attributes."""
+        return self.relation.data[self.relation.app].get("attributes", default=None)
+
+
+class CredentialsChangedEvent(S3Event):
+    """Event emitted when S3 credential are used on this relation."""
+
+
+class CredentialsGoneEvent(RelationEvent):
+    """Event emitted when S3 credential are removed from this relation."""
+
+
+class S3CredentialRequiresEvents(ObjectEvents):
+    """Event descriptor for events raised by the S3Provider."""
+
+    credentials_changed = EventSource(CredentialsChangedEvent)
+    credentials_gone = EventSource(CredentialsGoneEvent)
+
+
+S3_REQUIRED_OPTIONS = ["access-key", "secret-key", "bucket"]
+
+
+class S3Consumer(Object):
```

Comment:
```suggestion
class S3Requirer(Object):
```
I am bad with lingo requirer=consumer right ? if so, best to change this to Requirer as thats what the rest of the team follows